browser/brave_rewards/add_funds_popup.cc

```diff
@@ -0,0 +1,562 @@
+                                                   ContentSetting setting) {
+  DCHECK(map);
+  DCHECK(host);
+  DCHECK(type != CONTENT_SETTINGS_TYPE_PLUGINS);
+
+  const GURL gurl(host);
+  if (setting == CONTENT_SETTING_DEFAULT ||
+      setting != map->GetContentSetting(gurl, GURL(), type, std::string())) {
+    map->SetContentSettingDefaultScope(gurl, GURL(), type, std::string(),
+                                       setting);
+  }
+}
+
+}  // namespace
+
+namespace brave_rewards {
+
+// Pass-through to PopupContentSettingsBase which is in an anonymous namespace.
+class AddFundsPopupContentSettings : public PopupContentSettingsBase {
+ public:
+  AddFundsPopupContentSettings(Profile* profile)
+      : PopupContentSettingsBase(profile) {}
+  DISALLOW_COPY_AND_ASSIGN(AddFundsPopupContentSettings);
+};
+
+AddFundsPopup::AddFundsPopup()
+    : add_funds_popup_(nullptr),
+      popup_content_settings_(nullptr),
+      rewards_service_(nullptr) {}
+
+AddFundsPopup::~AddFundsPopup() {
+  ClosePopup();
+}
+
+// Show existing or open a new popup.
+void AddFundsPopup::ShowPopup(content::WebContents* initiator,
+                              RewardsService* rewards_service) {
+  add_funds_popup_ ? Focus() : OpenPopup(initiator, rewards_service);
+}
+
+void AddFundsPopup::OpenPopup(content::WebContents* initiator,
+                              RewardsService* rewards_service) {
+  DCHECK(!add_funds_popup_);
+  DCHECK(initiator);
+  DCHECK(rewards_service);
+  if (!initiator || !rewards_service)
+    return;
+
+  const std::map<std::string, std::string> addresses =
+      rewards_service->GetAddresses();
+  if (addresses.empty())
+    return;
+
+  content::WebContentsDelegate* wc_delegate = initiator->GetDelegate();
+  if (!wc_delegate)
+    return;
+
+  const GURL gurl(kAddFundsUrl);
+  const GURL referrer_gurl(std::string(brave_toolbar::kInternalUIScheme) +
+                           kRewardsHost);
+  const content::Referrer referrer(referrer_gurl,
+                                   blink::kWebReferrerPolicyAlways);
+  content::OpenURLParams params(gurl, referrer,
+                                WindowOpenDisposition::NEW_POPUP,
+                                ui::PAGE_TRANSITION_LINK, true);
+
+  // Supply addresses via post data. The data is currently in the query string
+  // format (application/x-www-form-urlencoded):
+  // addresses=UrlEscapedBase64EncodedStringifiedJSON.
+  params.uses_post = true;
+  const std::string data = ToQueryString(GetAddressesAsJSON(addresses));
+  params.post_data =
+      network::ResourceRequestBody::CreateFromBytes(data.data(), data.size());
+  params.extra_headers =
+      std::string("Content-Type: application/x-www-form-urlencoded\r\n") +
+      "Content-Length: " + std::to_string(data.size()) + "\r\n\r\n";
+
+  // Let popup content bypass shields, use camera and autoplay.
+  std::unique_ptr<AddFundsPopupContentSettings> popup_content_settings =
+      EnsureContentPermissions(initiator);
```

Comment:
can this validate the URL loaded in the initiator WebContents before relaxing the permissions? for instance if the webcontents are navigated away from the expected URL, the permissions should be removed (or never applied in the first place)

Reply:
If the webcontents is navigated away from, then, since this particular webcontents is a webui (for brave://rewards), the webui will be destroyed, which would cause this object (AddFundsPopup) to be destroyed as well and the permissions would be reset back in the destructor. Since both creation and deletion are initiated on the same thread I don't believe we can end up in a half-way state.

Reply:
sgtm